x-pack/plugins/security/common/licensing/license_service.ts

```diff
@@ -4,25 +4,43 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
+import { Observable, Subscription, BehaviorSubject } from 'rxjs';
 import { deepFreeze } from '../../../../../src/core/utils';
-import { ILicense } from '../../../licensing/server';
+import { ILicense } from '../../../licensing/common/types';
 import { SecurityLicenseFeatures } from './license_features';
 
 export interface SecurityLicense {
+  getChanges$(): Observable<void>;
   isEnabled(): boolean;
   getFeatures(): SecurityLicenseFeatures;
 }
 
+interface SetupDeps {
+  license$: Observable<ILicense>;
+}
+
 export class SecurityLicenseService {
-  public setup() {
+  private licenseSubscription?: Subscription;
+
+  public setup({ license$ }: SetupDeps) {
     let rawLicense: Readonly<ILicense> | undefined;
 
-    return {
-      update(newRawLicense: Readonly<ILicense>) {
-        rawLicense = newRawLicense;
-      },
+    const changesSubject$ = new BehaviorSubject<void>(undefined);
+
+    this.licenseSubscription = license$.subscribe(nextRawLicense => {
+      rawLicense = nextRawLicense;
+      changesSubject$.next();
```

Comment:
I feel like there must be a more elegant way to do this, but I'm not familiar enough with RxJS to know what that is...open to suggestions here.

I tried having `getChanges$()` below return `license$.pipe(switchMap(Rx.of()))`, but that wasn't behaving the way I thought it would (it never seemed to emit).

Reply:
> but I'm not familiar enough with RxJS to know what that is...

Yeah, me neither :see_no_evil: Right now I'm struggling with the question whether we should treat value stream (e.g. `license$` from Licensing plugin) and change stream (e.g. `changes$` on our license) in a same way. Technically the former should always eventually emit something whenever you subscribe to it, but the latter isn't since it reflects only changes from the current state. If we decide to separate these two notions then we should use `Subject` since `BehaviorSubject` will always emit value whenever you subscribe to it.

```ts
const changes$ = new Subject<void>();
this.licenseSubscription = license$
  .pipe(
    tap(nextRawLicense => (rawLicense = nextRawLicense)),
    map(() => undefined)
  )
  .subscribe(changes$);
.......
return {
  license: deepFreeze({
    changes$: changes$.asObservable(),
    ...
  }),
  ....
};
```

But at the same time it'd make consumer code that wants to react to changes a little bit more complex since now you need to extract common code so that you can call it for the license we have right now and for every its change, or use something like this:
```ts
concat(of(), license.changes$).pipe(
  map(() => license.getFeatures())
).subscribe(.....)
```

Which is basically equivalent to having both `license` and `license$` that would be used for different use cases ....

To summarize: 
* I think we need to figure out the common and good enough approach we'd use across plugins we own
* We don't need to spend time on this in this PR, we can move forward with what you have right now and once we have a critical mass of similar cases we can step back and refactor.

What do you think?

Reply:
Not entirely sure what the purpose of `changes$` is here, but it feels like it's always more valuable to share when something changed at the same time as sharing what changed. So instead of `changes$` emitting empty values, why not just emit the latest license?

The license plugin already only emits when licence values change https://github.com/elastic/kibana/blob/dce893bb79ee69f32cb20e105fa21700e827f2b2/x-pack/plugins/licensing/common/license_update.ts#L35 so `changes$` would be the same as `licence$` apart from omitting the initial value that `licence$` emits:
```ts
const changes$ = licence$.pipe(skip(1));
```

Reply:
> Technically the former should always eventually emit something whenever you subscribe to it, but the latter isn't since it reflects only changes from the current state. If we decide to separate these two notions then we should use Subject since BehaviorSubject will always emit value whenever you subscribe to it.

Yeah, I intentionally went with `BehaviorSubject` because it was easier to consume, but you're absolutely right, it's not technically correct that you'll receive an old value immediately after subscribing to something called `changes$`.

> Not entirely sure what the purpose of changes$ is here, but it feels like it's always more valuable to share when something changed at the same time as sharing what changed. So instead of changes$ emitting empty values, why not just emit the latest license?

In our case, the latest license is a frozen object which just contains a couple of functions. The functions themselves don't change when the license changes, but the output of those functions do. Emitting the latest license here would return the same thing you had before.


We could alternatively update each of those functions to return an observable, or supply a pair of functions: e.g. `getFeatures()` and `getFeatures$()`